store.go

```diff
@@ -748,6 +752,10 @@ func (s *store) GraphRoot() string {
 	return s.graphRoot
 }
 
+func (s *store) TransientStore() bool {
```

Comment:
> High-level:
> 
> Overall this feels like _quite a lot_ of complexity, and it seems fairly likely to be a maintenance headache:
> 
>     * Users must manually clean up half of the storage on reboot

I don't think this is actually a real problem. The intent is to ship podman with some systemd unit file that just does this for you. Then all you need to worry about is ordering that before starting any systemd services using podman.

>     * Where is the the lock file stored, in the persistent or non-persistent data?

It is important to get this right (i.e use the same in all places), and I think it makes sense to have it on the transient location as persistent locks have no meaning. However, I don't see any higher-level issue with this?

>     * IIRC _in general_, creating cross-store layers is unexpected; AFAIK that only works for overlay.

This however seems like a real issue. I'm pretty sure vfs is easy to make work, but for devmapper and btrfs this is not really possible. We could make it possible for the driver to expose a flag for supporting this or not, and then we could do the splitting of the layers store optionally.

>     * How are we going to keep this working? Do we need a whole new third mode for running all tests?

That seems like the "easiest" way, but I haven't yet looked at the storage testsuite. Is long test runtimes a problem for c/s currently?
 
> As a thought experiment, what would happen if
> 
>     * there were no separate transient mode
> 
>     * the “fast `podman run`” case was instead using a fresh graphroot in a temporary location, not even necessarily `tmpfs`, pointing at the long-term layer+image store using the existing “additional stores” mechanism?
> 
>     * there were some option to just turn off the `fdatasync`, so that all the metadata I/O has negligible latency

What would generally happen is that you disconnect the new container from all the global podman state. Like, unless you jump through hoops you can't: get logs for it, see it in podman ps, use podman health-checks, use named volumes, share namespaces with other containers, etc, etc. I'm also unsure how well podman handles the "read-only additional store" changing over time, although I see read locks for it, so that may be ok. (I guess some of these issues are due to the related boltdb transient store changes and not necessarily the storage changes, but we want a complete solution so they need to be considered together.)

The initial idea for transient mode was actually similar to this, i.e. just point rootdir and runroot somewhere on tmpfs, or mount a tmpfs on rootdir. Having a "no fsync" mode for storage is quite similar to this. 

The core issue here is that there is a mix of things in storage, some that we want to persist (backing store for image layers, image db, backing store for volumes, etc), and some we don't. The databases mix information about these, so any update has to be fully robust, even when just changing information about transient stuff. Splitting this up is really where the complexity of this PR comes from.

I don't really see a way to avoid the splitting though, because fundamentally we *do* want to have some persisted state, and we *do* want to use that global state with the transient state. However, maybe we can split it in a different way. For example, we could have a shared `/var/lib/containers/storage/overlay-layers/` directory, but we can have two json files in it, one for transient layers and one for permanent. That may make the changes less complex. Of course this also makes it harder to clean up to a "fresh" state in case of an unclean shutdown or other issue.

I'll have a look at the two-json-files approach and see how it would look.

> ? I think the major difference would be `podman run` triggering an image pull, because the image would be pulled into the temporary layer store, not the primary one. Is that a problem?

For our particular usecase, I don't think this is a problem, because we expect the install of an image to be separate from the running of it, but that depends on the user though.

Reply:
This is used by "podman info".

Reply:
That’s fair, filed bugs / support requests would benefit from having that information.

---

Maybe, instead / also, `DirectoriesToCleanUpOnReboot` (and have that on-boot systemd unit read this field instead of hard-coding knowledge)?

Reply:
Yes making this explicit makes sense. Would we specify a unit file to run and do a podman info and remove the directory?

Reply:
Or should there be a podman command to clean these up>\?

Reply:
Yeah, this is more complex than a fixed set of dirs, so probably would need a podman command.